p4-16/spec/P4-16-spec.mdk

```diff
@@ -2363,6 +2363,134 @@ struct Parsed_headers {
+of type `Fields`.  The first four fields of `Fields` are individual
+bytes of `Addresses.source`, and the field `Fields.all` is the
+concatenation of `Source.source` and `Source.destination`.
+
+A variable of type `Addresses` named `a` is instantiated, and an
+overlay named `q` is created using the `sourceQuads` overlay.  Now
+fields of the `a` structure can be accessed read-write using fields of
+the `q` structure.
+
+~ Begin P4Example
+struct Addresses {
+    bit<32> source;
+    bit<32> destination;
+}
+
+struct Fields {
+    bit<8> q0;
+    bit<8> q1;
+    bit<8> q2;
+    bit<8> q3;
+    bit<64> all;
+}
+
+overlay Fields sourceQuads(Addresses add) {
+   q0 = add.source[31:24];
+   q1 = add.source[23:16];
+   q2 = add.source[15:8];
+   q3 = add.source[7:0];
+   all = add.source ++ add.destination;
+}
+
+Addresses a = { 0x01234567, 0x89ABCDEF };
+Fields    q = sourceQuads(a);  // q is { 0x01, 0x23, 0x45, 0x67, 0x0123456789ABCDEF };
+a.source = 0;   // q is { 0x00, 0x00, 0x00, 0x00, 0x0000000089ABCDEF }, a is { 0x0, 0x89ABCDEF }
+a.source[7:0] = a.destination[7:0]; // q is { 0x00, 0x00, 0x00, 0xEF, 0x000000EF89ABCDEF }, a is { 0x000000EF, 0x89ABCDEF }
+~ End P4Example
+
+An overlay must define all fields of the destination structure.
+
+Overlays can only be applied to `struct` and `header` objects, and
+they can only produce `struct` objects.  A struct produced by an
+overlay can only contain fields of the base types: `bit<>` and
+`int<>`.  The only statement allowed within an overlay function is an
+assignment statement that assigns to a field of the destination.  The
+only expressions allowed within an overlay are field access, bit
+slicing and concatenation.  Referring to `header_union` fields within
+an overlay is not allowed.
+
+Under some restrictions overlays can be left-values:
+- they must be obtained by applying an overlay function to a left-value;
+- each bit from a source field must be used at most once.
+
+The second restriction is necessary to compute how the fields of the
+source structure are changed when the destination structure is
+updated.  In the previous example `q` is not a left value.  Updating
+an overlay which is a left value will update the original structure.
+If the input of an overlay is a header the rules about accesing
+uninitialized header fields are valid for the source header.
+
+~ Begin P4Example
+struct Addresses {
+    bit<32> source;
+    bit<32> destination;
+}
+
+struct Bytes {
+    bit<8> q0;
+    bit<8> q1;
+    bit<8> q2;
+    bit<8> q3;
+}
+
+overlay Bytes sourceQuads(Addresses add) {
+   q0 = add.source[31:24];
+   q1 = add.source[23:16];
+   q2 = add.source[15:8];
+   q3 = add.source[7:0];
+}
+
+Addresses a = { 0x01234567, 0x89ABCDEF };
+Fields    q = sourceQuads(a);  // q is { 0x01, 0x23, 0x45, 0x67 };
+q.q0 = 0; // q is { 0x0, 0x23, 0x45, 0x67 }, a is { 0x00234556, 0x89ABCDEF }
+~ End P4Example
+
+Overlays can be "stacked", as in the following example, where both `s`
+and `b` are left-values.
```

Comment:
Is the plan to simplify this proposal by disallowing "stacked" overlays?  If not, note that if both stacking and passing multiple parameters to an overlay function are allowed, it makes the compiler's job of determining whether an overlay can be used as an L-value a little bit more interesting: https://github.com/jafingerhut/p4-guide/tree/master/overlays#stacking-overlays-and-determining-whether-an-overlay-alias-can-be-an-l-value

Reply:
I think the trouble is from the multiple arguments; consider the case of calling an overlay with the same argument twice: o(a,a). But I think we should start simple and grow, so I will remove the stacking for now as well until someone asks for it.